mm2src/coins/utxo/rpc_clients/electrum_rpc/client.rs

```diff
@@ -306,10 +307,14 @@ impl ElectrumClient {
         let req_id = request.rpc_id();
         let request = json::to_string(&request).map_err(|e| JsonRpcErrorType::InvalidRequest(e.to_string()))?;
 
-        let connections = self.connection_manager.get_active_connections().await;
-        if connections.is_empty() {
-            return Err(JsonRpcErrorType::Transport("No connections available".to_string()));
-        }
+        let mut connections = vec![];
+        for _ in 0..10 {
+            connections = self.connection_manager.get_active_connections().await;
+            if !connections.is_empty() {
+                break;
+            }
+            Timer::sleep(0.5).await;
+        };
```

Comment:
I have no problem in making this forever waiting, it's actually the right solution IMHO (in case of internet disconnection for instance). For the tests we can use `#[cfg(test)]` to make this timeout.

BTW, if the electrum we are connected to is a faulty one (which has happened a lot before) can lead to `electrum_request_multi` failing here https://github.com/KomodoPlatform/komodo-defi-framework/blob/bf01ba0969835e9788ed7e7c3a5e8231b8a02053/mm2src/coins/utxo/rpc_clients/electrum_rpc/client.rs#L334 when other electrums can return a result.

Reply:
> BTW, for selective option, if the electrum we are connected to is a faulty one (which has happened a lot before) can lead to electrum_request_multi failing here 

aha i see, will wrap the whole thing in successive retries then.
but won't we ever have a request that is supposed to return an error in some case? meaning that erroring is the right response and has some valid interpretation (e.g. looking for txid that doesn't exist, if this errors then we will get stuck here forever).

Reply:
> BTW, for selective option, if the electrum we are connected to is a faulty one (which has happened a lot before) can lead to electrum_request_multi failing here

I believe this issue is not fully resolved, right? we still send the request in `electrum_request_multi` to only one electrum for the selective option. We should send it to all electrum sequentially, so for the selective option if the connected electrum / active connection returns an error here https://github.com/KomodoPlatform/komodo-defi-framework/blob/484143a6fabc816fb74d378725f244fe611a7f2d/mm2src/coins/utxo/rpc_clients/electrum_rpc/client.rs#L349-L352 we should try others sequentially by connecting and sending the request then disconnecting or using the one that got us a result as the active connection (similar to the rotation option we had before for multiple connections).

Reply:
> (similar to the rotation option we had before for multiple connections).

About this, was this removed or the rotation handled elsewhere?

Reply:
> we should try others sequentially by connecting and sending the request then disconnecting or using the one that got us a result as the active connection (similar to the rotation option we had before for multiple connections).

The next selections should be the task of the manager. When i wrote this i thought upon error the erroring connection will disconnect and a new one will be selected by the manager. Looks like this is not the case though. I can make `electrum_request` calls disconnect if not resolved successfully, this way we will have a new connection the next loop.

Reply:
Regarding my previous comment,

> aha i see, will wrap the whole thing in successive retries then.
> but won't we ever have a request that is supposed to return an error in some case? meaning that erroring is the right response and has some valid interpretation (e.g. looking for txid that doesn't exist, if this errors then we will get stuck here forever).

> UPD: so json-rpc errors are returned as valid responses (https://www.jsonrpc.org/specification#error_object). aside from one electrum version (1.0, which I think we don't use) that behaves badly on this request (https://electrumx.readthedocs.io/en/latest/protocol-methods.html#blockchain-transaction-broadcast), we shouldn't have any problems.

Turns out we internally convert json-rpc errorous responses into errors internally: https://github.com/KomodoPlatform/komodo-defi-framework/blob/484143a6fabc816fb74d378725f244fe611a7f2d/mm2src/common/jsonrpc_client.rs#L434-L436

So I am not entirely right about the claim above. The `JsonRpcErrorType::Response` error will need to be propagated I believe. And we have no way of knowing whether this is a `request problem` or a `server problem`. Maybe adding some measure to make sure all servers return the same error then propagate it (can't wait forever in this case since it's most likey a `request problem`).